File: src/engine/phaseMachine.ts
// src/engine/phaseMachine.ts — pure phase/state-machine primitive for game flow.
// A PhaseMachine is a plain, JSON-serializable description of allowed phase transitions.
// All functions are TOTAL: unknown phases / illegal transitions are no-ops, never throws.

/** A single phase in the machine: its id and the phases it may transition to. */
export interface PhaseNode<P extends string> {
  id: P;
  /** Phases reachable from this node. Empty for sinks. */
  to: P[];
  /** When true, this is an end state (a convenience flag; reachability is via `to`). */
  terminal?: boolean;
}

/** A declarative phase machine: a starting phase plus a node per phase id. */
export interface PhaseMachine<P extends string> {
  initial: P;
  nodes: Record<P, PhaseNode<P>>;
}

/** Identity helper that exists purely to infer/lock the literal phase type `P`. */
export function defineMachine<P extends string>(m: PhaseMachine<P>): PhaseMachine<P> {
  return m;
}

/**
 * True iff `from` is a known node whose `to` list contains `to`.
 * Unknown `from` (no node) or a `to` not listed yields false. Never throws.
 */
export function canGo<P extends string>(m: PhaseMachine<P>, from: P, to: P): boolean {
  const node = m.nodes[from];
  if (!node) return false;
  return node.to.indexOf(to) !== -1;
}

/**
 * Returns `to` when the transition is legal (see canGo), otherwise returns `from` unchanged.
 * Total: never throws on unknown or illegal transitions.
 */
export function go<P extends string>(m: PhaseMachine<P>, from: P, to: P): P {
  return canGo(m, from, to) ? to : from;
}

/**
 * True iff `phase` is a known node explicitly flagged `terminal`.
 * Unknown phases and unflagged nodes yield false. Never throws.
 */
export function isTerminal<P extends string>(m: PhaseMachine<P>, phase: P): boolean {
  const node = m.nodes[phase];
  return node ? node.terminal === true : false;
}

File: android-native/app/src/main/java/com/gamenight/party/engine/PhaseMachine.kt
package com.gamenight.party.engine

/**
 * Port of src/engine/phaseMachine.ts — a pure phase/state-machine primitive for game flow.
 *
 * A [PhaseMachine] is a plain, immutable description of allowed phase transitions. All functions are
 * TOTAL: unknown phases / illegal transitions are no-ops, never throwing. The phase type [P] is
 * typically an enum or a [String] (the web used string-literal unions).
 */

/** A single phase in the machine: its id, the phases it may transition to, and a terminal flag. */
data class PhaseNode<P>(
    val id: P,
    /** Phases reachable from this node. Empty for sinks. */
    val to: List<P>,
    /** When true, this is an end state (a convenience flag; reachability is via [to]). */
    val terminal: Boolean = false,
)

/** A declarative phase machine: a starting phase plus a node per phase id. */
data class PhaseMachine<P>(
    val initial: P,
    val nodes: Map<P, PhaseNode<P>>,
)

/** Identity helper that exists purely to mirror the web call sites. Returns [m] unchanged. */
fun <P> defineMachine(m: PhaseMachine<P>): PhaseMachine<P> = m

/**
 * True iff [from] is a known node whose `to` list contains [to]. Unknown [from] or a [to] not
 * listed yields false. Never throws.
 */
fun <P> canGo(m: PhaseMachine<P>, from: P, to: P): Boolean {
    val node = m.nodes[from] ?: return false
    return node.to.contains(to)
}

/**
 * Returns [to] when the transition is legal (see [canGo]), otherwise returns [from] unchanged.
 * Total: never throws on unknown or illegal transitions.
 */
fun <P> go(m: PhaseMachine<P>, from: P, to: P): P = if (canGo(m, from, to)) to else from

/**
 * True iff [phase] is a known node explicitly flagged terminal. Unknown phases and unflagged nodes
 * yield false. Never throws.
 */
fun <P> isTerminal(m: PhaseMachine<P>, phase: P): Boolean {
    val node = m.nodes[phase] ?: return false
    return node.terminal
}
